src/network/protocol/messages/block.zig

```diff
@@ -0,0 +1,201 @@
+            txn.deinit();
+        }
+        self.txns.deinit();
+    }
+
+    /// Serialize the message as bytes and write them to the Writer.
+    ///
+    /// `w` should be a valid `Writer`.
+    pub fn serializeToWriter(self: *const Self, w: anytype) !void {
+        comptime {
+            if (!std.meta.hasFn(@TypeOf(w), "writeInt")) @compileError("Expects r to have fn 'writeInt'.");
+            if (!std.meta.hasFn(@TypeOf(w), "writeAll")) @compileError("Expects r to have fn 'writeAll'.");
+        }
+
+        try w.writeInt(i32, self.block_header.version, .little);
+        try w.writeAll(&self.block_header.prev_block);
+        try w.writeAll(&self.block_header.merkle_root);
+        try w.writeInt(i32, self.block_header.timestamp, .little);
+        try w.writeInt(i32, self.block_header.nbits, .little);
+        try w.writeInt(i32, self.block_header.nonce, .little);
+
+        try self.txn_count.encodeToWriter(w);
+
+        for (self.txns.items) |txn| {
+            try txn.serializeToWriter(w);
+        }
+    }
+
+    /// Serialize a message as bytes and write them to the buffer.
+    ///
+    /// buffer.len must be >= than self.hintSerializedLen()
+    pub fn serializeToSlice(self: *const Self, buffer: []u8) !void {
+        var fbs = std.io.fixedBufferStream(buffer);
+        try self.serializeToWriter(fbs.writer());
+    }
+
+    /// Serialize a message as bytes and return them.
+    pub fn serialize(self: *const BlockMessage, allocator: std.mem.Allocator) ![]u8 {
+        const serialized_len = self.hintSerializedLen();
+
+        const ret = try allocator.alloc(u8, serialized_len);
+        errdefer allocator.free(ret);
+
+        try self.serializeToSlice(ret);
+
+        return ret;
+    }
+
+    pub fn deserializeReader(allocator: std.mem.Allocator, r: anytype) !BlockMessage {
+        comptime {
+            if (!std.meta.hasFn(@TypeOf(r), "readInt")) @compileError("Expects r to have fn 'readInt'.");
+            if (!std.meta.hasFn(@TypeOf(r), "readNoEof")) @compileError("Expects r to have fn 'readNoEof'.");
+            if (!std.meta.hasFn(@TypeOf(r), "readAll")) @compileError("Expects r to have fn 'readAll'.");
+            if (!std.meta.hasFn(@TypeOf(r), "readByte")) @compileError("Expects r to have fn 'readByte'.");
+        }
+
+        var bm: Self = undefined;
+
+        bm.block_header.version = try r.readInt(i32, .little);
+
+        const prev_block_raw_bytes = try read_bytes_exact(allocator, r, 32);
+        defer allocator.free(prev_block_raw_bytes);
+        @memcpy(&bm.block_header.prev_block, prev_block_raw_bytes);
+
+        const merkle_root_raw_bytes = try read_bytes_exact(allocator, r, 32);
+        defer allocator.free(merkle_root_raw_bytes);
+        @memcpy(&bm.block_header.merkle_root, merkle_root_raw_bytes);
+
+        bm.block_header.timestamp = try r.readInt(i32, .little);
+        bm.block_header.nbits = try r.readInt(i32, .little);
+        bm.block_header.nonce = try r.readInt(i32, .little);
```

Comment:
Coudn't this be a `pub fn deserializeReader(allocator: std.mem.Allocator, r: anytype) !Blockheader`? Implemented in the BlockHeader scope